### PR TITLE
Multi dimensional output

### PR DESCRIPTION
**Pull request by kxk302**
### Pull Request Checklist

- [ X] Check python style, using `flake8 ./galaxy_ml`.
- [ X] Run API tests, using `cd galaxy_ml; nosetests ./tests -sv`.
- [ ] Check tool lint, using `cd galaxy_ml/tools; planemo lint`.
- [ ] Run tool tests, using `cd galaxy_ml/tools; planemo test`.
- [ ] Update `docs/CHANGELOG`, if applicable.
- [ X] Update `README.md`, if applicable.

Fix an issue that disallowed multi dimensional output files. Before the fix, the output file could only be a vector (Many rows, 1 column). Now, the output file can be a matrix (many rows, many columns). This is needed to solve multiclass/multilabel classification problems.

Wrote a failing test to re-produce the problem. The fix caused the test to pass. Updated README to document how to run the tests and generate an HTML report. Updated the .gitignore to ignore temporary test/build files.

### CODE REVIEW

**Comment by kxk302** [comment]
@qiagu This PR should fix the issue we discussed. Thanks for your help.

File: README.md [hunk 1]
```diff
@@ -93,6 +93,29 @@ python setup.py build_ext --inplace
...
+To run all tests and generate an HTML report:
+```
+nosetests ./tests --with-html --html-file=./report.html
+```
+
+To run tests in a specific file (e.g., test_keras_galaxy.py file) and generate an HTML report
+nosetests ./tests/test_keras_galaxy.py --with-html --html-file=./report.html
+
+To run a specific test in a specific file (e.g., test_multi_dimensional_output test in test_keras_galaxy.py file) and generate an HTML report
+nosetests ./tests/test_keras_galaxy.py:test_multi_dimensional_output --with-html --html-file=./report.html
```
Comment (qiagu) [inline README.md:119]: Thanks for making this!

File: galaxy_ml/keras_galaxy_models.py [hunk 2]
```diff
@@ -876,7 +876,12 @@ def fit(self, X, y, class_weight=None, **kwargs):
...
         X : array-like, shape `(n_samples, feature_arrays)`
         """
-        X, y = check_X_y(X, y, accept_sparse=['csr', 'csc'], allow_nd=True)
+        multi_output = False
+        if y.ndim == 2:
+            multi_output = True
+
+        X, y = check_X_y(X, y, accept_sparse=['csr', 'csc'], allow_nd=True,
+                         multi_output=multi_output)
         check_classification_targets(y)
```
Comment (qiagu) [inline galaxy_ml/keras_galaxy_models.py:885]: Can we just let `multi_output=True` for this? If the type of classification problem needs to be checked, sklean provides couple of utils to call, such as `type_of_target`, 'is_multilabel' and so on.
https://github.com/scikit-learn/scikit-learn/blob/8ea176ae0ca535cdbfad7413322bbc3e54979e4d/sklearn/utils/multiclass.py#L186
https://github.com/scikit-learn/scikit-learn/blob/8ea176ae0ca535cdbfad7413322bbc3e54979e4d/sklearn/utils/multiclass.py#L113

Comment (qiagu) [reply galaxy_ml/keras_galaxy_models.py:885]: Also, there is another `check_X_y` under the `KerasGBatchClassifier`. Could you please change that as well?

Comment (kxk302) [reply galaxy_ml/keras_galaxy_models.py:885]: Will change that one too.

Comment (kxk302) [reply galaxy_ml/keras_galaxy_models.py:885]: > Can we just let `multi_output=True` for this? If the type of classification problem needs to be checked, sklean provides couple of utils to call, such as `type_of_target`, `is_multilabel` and so on.
> https://github.com/scikit-learn/scikit-learn/blob/8ea176ae0ca535cdbfad7413322bbc3e54979e4d/sklearn/utils/multiclass.py#L186
> https://github.com/scikit-learn/scikit-learn/blob/8ea176ae0ca535cdbfad7413322bbc3e54979e4d/sklearn/utils/multiclass.py#L113

What I did was to make sure the existing code work as before (Flag remains False), and only change the flag if output data is multidimensional (New use case).

**Comment by kxk302** [comment]
@qiagu pushed the revised code.

**Comment by qiagu** [comment]
@kxk302  You said some were pushed here. However, I didn't see any new commit coming. 
Also, I have a PR against your branch. Could you please have a look?

File: galaxy_ml/keras_galaxy_models.py [hunk 3]
```diff
@@ -1104,7 +1109,12 @@ def fit(self, X, y=None, class_weight=None, sample_weight=None, **kwargs):
         self.data_generator_.set_processing_attrs()
 
         if y is not None:
-            X, y = check_X_y(X, y, accept_sparse=['csr', 'csc'], allow_nd=True)
+            multi_output = False
```
Comment (kxk302) [inline galaxy_ml/keras_galaxy_models.py:1112]: @qiagu This is the change I pushed. Added the fix to the other method too, as you suggested.

Comment (qiagu) [reply galaxy_ml/keras_galaxy_models.py:1112]: Okay. Thanks.

**Comment by qiagu** [comment]
@kxk302  Merged. Thanks for making this PR!

**Comment by kxk302** [comment]
> @kxk302 Merged. Thanks for making this PR!

Thank you @qiagu for your quick review! Just wondering, what is the process for deploying this to Galaxy.eu? Should I contact their admins to have this fix deployed? Appreciate any info on the deployment process.

**Comment by qiagu** [comment]
@kxk302  It looks we still need scorer for multi-class classification problem.
https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/keras_galaxy_models.py#L1345

**Comment by kxk302** [comment]
> @kxk302 It looks we still need scorer for multi-class classification problem.
> https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/keras_galaxy_models.py#L1345

@qiagu I just created another PR (https://github.com/goeckslab/Galaxy-ML/pull/56) to fix some README formatting issues. I also added a confusion matrix to multi-class classification test and it seems to work fine. How does the scorer for multi-class classification get invoked?

**Comment by qiagu** [comment]
@kxk302  Right now, the `evaluate` method is only called in the `_fit_and_score` within sklearn's searchCV. I'm not clear how to implement the multi-class scoring yet. It seems if you don't want to use keras model in sklearn's searchcv or use sklearn's scorers in keras, current implement is fine.

**Comment by qiagu** [comment]
>Just wondering, what is the process for deploying this to Galaxy.eu? Should I contact their admins to have this fix deployed?

Sorry for missing this. It's usually a long process for tools to get deployed. First, publish a new verion of Galaxy-ML, like v0.8.3, then build corresponding conda environment in the bioconda channel, after that, update the tools (probably dependencies only) in Bjorn's repo. Bjorn takes care of the rest, I guess.

**Comment by kxk302** [comment]
> @kxk302 Right now, the `evaluate` method is only called in the `_fit_and_score` within sklearn's searchCV. I'm not clear how to implement the multi-class scoring yet. It seems if you don't want to use keras model in sklearn's searchcv or use sklearn's scorers in keras, current implement is fine.

I am not using sklearn's searchCV right now, so i guess the current implementation is fine, and we can address the evaluate issue separately.

**Pull request merged**